src/Schemes/AffineSchemes/Objects/Constructors.jl

```diff
@@ -0,0 +1,670 @@
+
+julia> R = OO(X)
+Multivariate Polynomial Ring in x1, x2, x3 over Rational Field
+
+julia> (x1,x2,x3) = gens(R)
+3-element Vector{fmpq_mpoly}:
+ x1
+ x2
+ x3
+
+julia> subscheme(X,ideal(R,[x1*x2]))
+Spec of Quotient of Multivariate Polynomial Ring in x1, x2, x3 over Rational Field by ideal(x1*x2)
+```
+"""
+function subscheme(X::AbsSpec, I::Ideal)
+  base_ring(I) == OO(X) || return subscheme(X, ideal(OO(X), OO(X).(gens(I)))) # this will throw if coercion is not possible
+  return Spec(quo(OO(X), I)[1])
+end
+
+
+
+########################################################
+# (6) Open subschemes
+########################################################
+
+@Markdown.doc """
+    hypersurface_complement(X::AbsSpec, f::RingElem)
+
+For a scheme ``X = Spec(R)`` and an element ``f ∈ R``
+this returns the open subscheme ``U = Spec(R[f⁻¹]) = X ∖ V(f)``
+defined by the complement of the vanishing locus of ``f`.
+
+# Examples
+```jldoctest
+julia> X = affine_space(QQ,3)
+Spec of Multivariate Polynomial Ring in x1, x2, x3 over Rational Field
+
+julia> R = OO(X)
+Multivariate Polynomial Ring in x1, x2, x3 over Rational Field
+
+julia> (x1, x2, x3) = gens(R)
+3-element Vector{fmpq_mpoly}:
+ x1
+ x2
+ x3
+
+julia> hypersurface_complement(X, x1)
+Spec of localization of Multivariate Polynomial Ring in x1, x2, x3 over Rational Field at the powers of fmpq_mpoly[x1]
+```
+"""
+function hypersurface_complement(X::AbsSpec, f::RingElem)
+  return hypersurface_complement(underlying_scheme(X), f)::AbsSpec
+end
+
+function hypersurface_complement(X::SpecType, f::RingElem) where {SpecType<:AbsSpec{<:Any, <:MPolyQuoLocalizedRing}}
+  parent(f) == OO(X) || return hypersurface_complement(X, OO(X)(f))
+  h = lifted_numerator(f)
+  U = MPolyPowersOfElement(h)
+  W, _ = Localization(OO(X), U)
+  return Spec(W)
+end
+
+function hypersurface_complement(X::SpecType, f::RingElem) where {SpecType<:AbsSpec{<:Any, <:MPolyLocalizedRing}}
+  parent(f) == OO(X) || return hypersurface_complement(X, OO(X)(f))
+  h = numerator(f)
+  U = MPolyPowersOfElement(h)
+  W, _ = Localization(OO(X), U)
+  return Spec(W)
+end
+
+function hypersurface_complement(X::SpecType, f::RingElem) where {SpecType<:AbsSpec{<:Any, <:MPolyRing}}
+  parent(f) == OO(X) || return hypersurface_complement(X, OO(X)(f))
+  U = MPolyPowersOfElement(f)
+  W, _ = Localization(OO(X), U)
+  return Spec(W)
+end
+
+function hypersurface_complement(X::SpecType, f::RingElem) where {SpecType<:AbsSpec{<:Any, <:MPolyQuo}}
+  parent(f) == OO(X) || return hypersurface_complement(X, OO(X)(f))
+  U = MPolyPowersOfElement(lift(f))
+  W, _ = Localization(OO(X), U)
+  return Spec(W)
+end
+
+
+@Markdown.doc """
+    hypersurface_complement(X::AbsSpec, f::Vector{<:RingElem})
+
+For a scheme ``X = Spec(R)`` and elements ``f1, f2, ... ∈ R``
+this returns the open subscheme ``U = Spec(R[f1⁻¹,f2⁻¹, ...]) = X ∖ V(f1, f2, ...)``
+defined by the complement of the vanishing locus ``V(f1, f2, ...)``.
```

Comment:
What is intended here is X \setminus (V(f1*f2*...), i.e. X \setminus (V(f1) \cup V(f2) \cup ...), not X \setminus (V(f1,f2,...)) = X\setminus (V(f1) \cap V(f2) ...). The syntax using a vector is only intended to pass on a known factorization to avoid repeatedly using factorization.

Reply:
We will talk this through and correct it. As it stands here now, it's wrong, but probably hard to guess from the code alone.

Reply:
Ok. Thank you. Let me change that.